Restore shutdown

sgranjoux opened the pull request:
Restore the ability to stop the current task on log out.

Add a new class to listen for GNOME session events on DBus.
Restore shutdown check box,

mwilck commented:
I haven't looked at details yet, but in general this is highly welcome. I wonder if you've also looked into stopping tracking when the lock screen is activated? Have you seen #568 and the related discussion?

sgranjoux commented:
I'm using hamster since a long time and I'm missing this 'stop tracking on shutdown` feature. But I haven't much time to work on this so it takes a quite long time to annoy me enough to take more active actions.

I have take a quick look at the 'stop tracking when idle' feature when writing this. It looks like it could be restored too but I haven't planned to do it, at least in a short future.

I have just read #568 and a few related discussions. Well, at least there are some life, we will see how it goes.

matthijskooijman commented:
I've had a look at this PR. I like the idea of restoring this behavior.

What confuses me, though, is that one commit talks about shutdown and the other about logout, but in practice I think the feature applies to both (i.e. any clean termination of the gnome session), right?

Also, it seems this PR is very Gnome-specific? IMHO it's acceptable to only support a subset of environments (i.e. only Gnome right now maybe), but what happens with the current code on e.g. KDE? Is there (I believe not?) any way for the user to know that the feature is not supported, or is the checkbox in the settings just shown but ineffective? That would not be acceptable IMHO. Also, would the lack of a GnomeSession DBUS service be handled gracefully?

I wonder if we cannot detect logout in a desktop-environment-agnostic way? If the hamster daemon is running inside the user session, it will be terminated as part of logout, so couldn't that just be a moment to stop tracking? Or is that hard to distinguish from manually stopping hamster (but does that ever happen)?

Alternatively, maybe systemd-logind can help here? That might work across different desktop environments and maybe even allow detecting any secondary sessions for the same user, if we think that's relevant. Though logind might not be enabled on all systems, of course.

sgranjoux commented:
Indeed, it stop the tracking on logout. I have kept shutdown because I think it was the name used before this feature has been removed.

I have not checked what's happen on KDE, I think it will just do nothing.

I think I have checked how this feature has been implemented and just replace it with something similar. I haven't found a general way to achieve that. I think it's fine if we stop the tracking when the user stop the daemon but I don't know if stopping the tracking won't need another user process which can be already stopped.

Then using systemd-logind or something else for KDE will probably not work on all systems, so in any case we need to check what's happen when the DBUS service is missing and probably disable this option.

GeraldJansen commented:
This definitely breaks Hamster on non-Gnome systems. I've tried on XFCE and get this vomit:

```
[hamster]$ hamster
hamster-service up
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 177, in activate_name_owner
    return self.get_name_owner(bus_name)
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 361, in get_name_owner
    return self.call_blocking(BUS_DAEMON_NAME, BUS_DAEMON_PATH,
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NameHasNoOwner: Could not get owner of name 'org.gnome.SessionManager': no such name

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/d/gj/hamster/src/hamster-service.py", line 467, in <module>
    storage = Storage(loop)
  File "/d/gj/hamster/src/hamster-service.py", line 55, in __init__
    DbusSessionListener.__init__(self)
  File "/d/gj/hamster/src/hamster/session.py", line 38, in __init__
    session_manager = session_bus.get_object("org.gnome.SessionManager",
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 241, in get_object
    return self.ProxyObjectClass(self, bus_name, object_path,
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 250, in __init__
    self._named_service = conn.activate_name_owner(bus_name)
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 182, in activate_name_owner
    self.start_service_by_name(bus_name)
  File "/usr/lib/python3/dist-packages/dbus/bus.py", line 277, in start_service_by_name
    return (True, self.call_blocking(BUS_DAEMON_NAME, BUS_DAEMON_PATH,
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name org.gnome.SessionManager was not provided by any .service files
ERROR:dbus.proxies:Introspect error on :1.522:/org/gnome/Hamster: dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
Traceback (most recent call last):
  File "/d/gj/hamster/src/hamster-cli.py", line 149, in on_activate_window
    self._open_window(action.get_name(), data)
  File "/d/gj/hamster/src/hamster-cli.py", line 184, in _open_window
    self.overview_controller = Overview()
  File "/d/gj/hamster/src/hamster/overview.py", line 477, in __init__
    self.find_facts()
  File "/d/gj/hamster/src/hamster/overview.py", line 531, in find_facts
    self.facts = self.storage.get_facts(start, end, search_terms=search)
  File "/d/gj/hamster/src/hamster/client.py", line 156, in get_facts
    for fact in self.conn.GetFactsJSON(dbus_range, search_terms)]
  File "/d/gj/hamster/src/hamster/client.py", line 99, in conn
    server_version = self._connection.Version()
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 72, in __call__
    return self._proxy_method(*args, **keywords)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.522 was not provided by any .service files
ERROR:dbus.proxies:Introspect error on :1.524:/org/gnome/Hamster: dbus.exceptions.DBusException: org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
Traceback (most recent call last):
  File "/d/gj/hamster/src/hamster-cli.py", line 149, in on_activate_window
    self._open_window(action.get_name(), data)
  File "/d/gj/hamster/src/hamster-cli.py", line 184, in _open_window
    self.overview_controller = Overview()
  File "/d/gj/hamster/src/hamster/overview.py", line 477, in __init__
    self.find_facts()
  File "/d/gj/hamster/src/hamster/overview.py", line 531, in find_facts
    self.facts = self.storage.get_facts(start, end, search_terms=search)
  File "/d/gj/hamster/src/hamster/client.py", line 156, in get_facts
    for fact in self.conn.GetFactsJSON(dbus_range, search_terms)]
  File "/d/gj/hamster/src/hamster/client.py", line 99, in conn
    server_version = self._connection.Version()
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 72, in __call__
    return self._proxy_method(*args, **keywords)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.524 was not provided by any .service files
```

sgranjoux commented:
Thank you for checking that, I have not much time but I will fix that.

sgranjoux commented:
I have caught the exception and added a warning if the GNOME session is not available. The I have added a tool tip on the setting to warn the user that this function is not available outside GNOME. It was pretty straightforward.

I have put all this in a third commit. Do you prefer that I merge those changes in the previous two commits?

I have kept the name shutdown because it was the old name. Do you prefer that I call that "Stop tracking on logout"?

Then for other environments. I'm afraid that stopping the tracking when the hamster daemon is terminated by the log out could lead to nasty bug depending on the order in which other processes are killed. Do you think it can be an issue?

Else, we can probably get an end session signal on KDE and perhaps XFCE or by using systemd. Do you have a preference?

GeraldJansen commented on src/hamster/session.py:
```diff
@@ -35,34 +37,36 @@ def __init__(self):
+                dbus_interface = "org.gnome.SessionManager.ClientPrivate",
+                bus_name = "org.gnome.SessionManager")
+        except dbus.exceptions.DBusException:
+            logger.warning("Unable to connect to GNOME session manager, stop trcking on logout won't work")
```
Please either make this logger.info() or just replace this line with `pass` so command line users don't see this message on every single command (because warning is the default logging level). Also trcking -> tracking.

sgranjoux replied:
Ok, I have changed it but I think this message will appear only when your start the hamster daemon so probably not on each command.

GeraldJansen replied:
Ah right, I forgot that I was using a cover script that restart the daemon before each command. Thanks anyway.

GeraldJansen commented on data/preferences.ui:
```diff
@@ -49,6 +49,8 @@
                         <property name="xalign">0.5</property>
                         <property name="draw_indicator">True</property>
+                        <property name="has_tooltip">True</property>
+                        <property name="tooltip_text" translatable="yes">Work only on GNOME</property>
```
"Only works on GNOME"

GeraldJansen replied:
Actually I think it would be better if the main description was changed to "Stop tracking on shutdown (GNOME only)".

sgranjoux replied:
The tooltip appears quite easily here so I think it was enough but I have changed the label.

GeraldJansen commented:
Tried again on XFCE and no longer seems to cause problems. I didn't notice any problem when killing the hamster daemon from the command line or when logging out (i.e. an ongoing activity remains open) . For me it is not a problem to introduce functionality available only on GNOME.

GeraldJansen commented:
> I wonder if we cannot detect logout in a desktop-environment-agnostic way?

Maybe this should be a new issue / feature request. Perhaps one could listen for [org.freedesktop.login1](https://www.freedesktop.org/software/systemd/man/org.freedesktop.login1.html) PrepareForShutdown and PrepareForSleep signals?

matthijskooijman commented on src/hamster/session.py:
```diff
@@ -0,0 +1,85 @@
+        """Inform that the session is about to end. It must reply with
+        EndSessionResponse within one second telling if it is ok to proceed
+        or not and why. If flags is true, the session will end anyway."""
+        self.__session_client_private_iface.EndSessionResponse(True, "")
```
Why is this handler needed? If we're not doing anything other than saying it is ok to end the session, isn't this just the default? Or is it so that *if* you register something (I'm not sure about right terminology here) against SessionManager, that you *need* to also handle this QueryEndSession event?

sgranjoux replied:
I have not checked, but according to the comment I have understood that when you register against the SessionManager you have to handle this message. 
If it's not done, I think it can delay the log out by 1 seconds, eventually the session manager could remove the corresponding program and we will not get the end session message.

matthijskooijman commented on src/hamster/session.py:
```diff
@@ -0,0 +1,85 @@
+        """Override this method to do something at the end of a session.
+        It must not attempt to interact with the user and will be given
+        a maximum of ten seconds to perform the actions."""
+        pass
```
Maybe this should raise `NotImplementedError`?

sgranjoux replied:
I think pass is fine, there is no need to do something.
Eventually, I could extend this object to provide a hook in query_end_session_handler. I will allow to use it to delay the end of a session and it could be that nothing is needed when the session end. Anyway, it's not done currently, so I could raise `NotImplementedError` if you think it's better.

matthijskooijman commented on src/hamster-service.py:
```diff
@@ -38,7 +40,7 @@
 
 
-class Storage(db.Storage, dbus.service.Object):
+class Storage(db.Storage, dbus.service.Object, DbusSessionListener):
```
I wonder if adding a mixin class here is the cleanest way to implement this. Also, it doesn't seem a responsibility of a class called Storage to implement a "stop-tracking-on-shutdown", but maybe that class already does more than its name implies (haven't looked too closely now).

sgranjoux replied:
I agree with that. It's probably the simplest way but not the cleanest one.

I think I can create a separate object having a reference on the storage object to call StopTracking. To make it even more separated, this object can send a stop tracking message on the dbus, it could be even a separate process.

matthijskooijman reviewed (commented):
I had a look over the code. It looks ok overall, left some inline remarks (mostly questions). I haven't tested the code.

> Maybe this should be a new issue / feature request. Perhaps one could listen for org.freedesktop.login1 PrepareForShutdown and PrepareForSleep signals?

Agreed, I do no think this must be solved before merging this, provided that this limitation if sufficiently clear for the user.

On that track, I like mentioning it is GNOME-only in the label more than the tooltip, since people might not be reading the tooltip and then be surprised it doesn't work.

> I have kept the name shutdown because it was the old name. Do you prefer that I call that "Stop tracking on logout"?

I personally would find "shutdown" confusing, since that implies a system shutdown to me.

Maybe "Stop tracking when GNOME session ends" could be an alternative, which neatly also includes the GNOME limitation?


> Then for other environments. I'm afraid that stopping the tracking when the hamster daemon is terminated by the log out could lead to nasty bug depending on the order in which other processes are killed. Do you think it can be an issue?

What issue are you afraid of exactly? I can imagine that whenever the hamster process is terminated, it could just stop tracking. This probably needs a SIGTERM handler or something like that, though maybe Python has some "on process end" code (or maybe the mainloop already neatly stops running on a SIGTERM and you could just add some code after the mainloop?).

I still think it would be good to consider this alternative, since if it would work, would not only work on non-GNOME systems, but also on GNOME systems, so I expect that the code for this would be simpler and maybe less fragile than listening on the gnome session dbus.

One thing that might prevent this from working properly, though, is that it relies on the hamster daemon being stopped when the login session ends, which might not actually be the case. I.e. there is systemd-logind, which I think supports processes that are tied to a user, not a particular login session (so they are killed only when the last login session of that user is logged out). I'm not sure how that works with dbus-autostart, though (which I think is how hamster is being started now in most cases? Or is it a .desktop file in an autostart directory?).

sgranjoux commented:
> Maybe "Stop tracking when GNOME session ends" could be an alternative, which neatly also includes the GNOME limitation?
That's fine for me. 

Should I change the name of the gconf key from `stop-on-shutdown` to `stop-on-logout`?


> What issue are you afraid of exactly?

When you log out, all user programs are killed, I think it includes supporting processes used to exchange dbus signal or read gconf key. I'm afraid that those functions are used by hamster and don't work at that time but I haven't checked anything. 

> One thing that might prevent this from working properly, though, is that it relies on the hamster daemon being stopped when the login session ends, which might not actually be the case. I.e. there is systemd-logind, which I think supports processes that are tied to a user, not a particular login session

I don't think it's a problem is the tracking is not stop in such case.
